raiden/network/transport/matrix/transport.py

```diff
@@ -955,79 +966,75 @@ def _get_room_for_address(
         address_hex = to_normalized_address(address)
         msg = f"address not health checked: me: {self._user_id}, peer: {address_hex}"
         assert address and self._address_mgr.is_address_known(address), msg
+        assert self._raiden_service is not None
 
-        # filter_private is done in _get_room_ids_for_address
-        room_ids = self._get_room_ids_for_address(address)
-        if room_ids:  # if we know any room for this user, use the first one
-            # This loop is used to ignore any broadcast rooms that may have 'polluted' the
-            # user's room cache due to bug #3765
-            # Can be removed after the next upgrade that switches to a new TokenNetworkRegistry
-            while room_ids:
-                room_id = room_ids.pop(0)
-                room = self._client.rooms[room_id]
-                if not self._is_broadcast_room(room):
-                    self.log.debug("Existing room", room=room, members=room.get_joined_members())
-                    return room
-                self.log.warning("Ignoring broadcast room for peer", room=room, peer=address_hex)
-
-        assert self._raiden_service is not None, "_raiden_service not set"
-        address_pair = sorted(
-            [to_normalized_address(address) for address in [address, self._raiden_service.address]]
-        )
-        room_name = make_room_alias(self.chain_id, *address_pair)
-
-        # no room with expected name => create one and invite peer
+        # filter peer_candidates
         peer_candidates = [
             self._get_user(user) for user in self._client.search_user_directory(address_hex)
         ]
-
-        # filter peer_candidates
         peers = [user for user in peer_candidates if validate_userid_signature(user) == address]
+
         if not peers and not allow_missing_peers:
             self.log.error("No valid peer found", peer_address=to_checksum_address(address))
             return None
 
+        # filter_private is done in _get_room_ids_for_address
+        room_ids = self._get_room_ids_for_address(address)
+        online_user_ids = self._address_mgr.get_online_user_ids_for_address(address)
+        while room_ids:
+            room = self._client.rooms[room_ids.pop()]
+            member_ids = {member.user_id for member in room.get_joined_members()}
+
+            if member_ids & online_user_ids:
+                self.log.debug(
+                    "Existing room",
+                    room=room,
+                    members=member_ids,
+                    listening_peers=member_ids & online_user_ids,
+                )
+                return room
+
+            # Handle the case where an empty room was found only after all other rooms we're tried.
+            # Re-invite the peers in case we got whitelisted in the meantime and return it.
+            elif not room_ids:
```

Comment:
This if would be true if the `room_ids` is empty, which is the same thing as the whole loop finishing. Please move this block of code out of the loop and remove the test.

Reply:
After a second look, I don't think you wanted to do `elif not room_ids`, maybe this should be `elif not member_ids`?

Reply:
See my comment below, the idea here was to return a room without online members only in the last iteration, which is why I used elif not room_ids after room_ids.pop() to avoid counting an index or something.

Reply:
Okay, after the call I understood what you meant, the algorithm is:

```
- if there are rooms:
  - iterator over the rooms and return the first one that our partner node is online
  - otherwise, return the last room if the partner is not online on any of these
- create a room
```
My suggestion is to change the code to:
```python
def get_best_room(rooms):
  for r in rooms:
    if partner_available_at(r): return r
  if rooms: return rooms[0]

def get_room_for():
  ...
  if rooms:
    return get_best_room(rooms)
  return create_room_invite_and_wait()